File: Backend/sms-backend/src/schemas/people/student.py
from datetime import date
from typing import Optional, List
from uuid import UUID
from pydantic import BaseModel

from src.schemas.auth.user import UserBase, UserCreate, UserUpdate, User


class StudentBase(UserBase):
    """Base schema for Student model."""
    admission_number: str
    roll_number: Optional[int] = None
    grade: Optional[str] = None
    section: Optional[str] = None
    admission_date: Optional[date] = None
    date_of_birth: Optional[date] = None
    gender: Optional[str] = None
    blood_group: Optional[str] = None
    nationality: Optional[str] = None
    religion: Optional[str] = None
    address: Optional[str] = None
    city: Optional[str] = None
    county: Optional[str] = None
    country: Optional[str] = None
    whatsapp_number: Optional[str] = None
    emergency_contact: Optional[str] = None
    photo: Optional[str] = None
    status: str = "active"


class StudentCreate(UserCreate):
    """Schema for creating a new student."""
    admission_number: Optional[str] = None
    roll_number: Optional[int] = None
    grade: Optional[str] = None
    section: Optional[str] = None
    admission_date: Optional[date] = None
    date_of_birth: Optional[date] = None
    gender: Optional[str] = None
    blood_group: Optional[str] = None
    nationality: Optional[str] = None
    religion: Optional[str] = None
    address: Optional[str] = None
    city: Optional[str] = None
    county: Optional[str] = None
    country: Optional[str] = None
    whatsapp_number: Optional[str] = None
    emergency_contact: Optional[str] = None
    photo: Optional[str] = None
    status: str = "active"


class StudentUpdate(UserUpdate):
    """Schema for updating a student."""
    admission_number: Optional[str] = None
    roll_number: Optional[int] = None
    grade: Optional[str] = None
    section: Optional[str] = None
    admission_date: Optional[date] = None
    date_of_birth: Optional[date] = None
    gender: Optional[str] = None
    blood_group: Optional[str] = None
    nationality: Optional[str] = None
    religion: Optional[str] = None
    address: Optional[str] = None
    city: Optional[str] = None
    county: Optional[str] = None
    country: Optional[str] = None
    whatsapp_number: Optional[str] = None
    emergency_contact: Optional[str] = None
    photo: Optional[str] = None
    status: Optional[str] = None
    exit_date: Optional[date] = None
    graduation_date: Optional[date] = None
    withdrawal_reason: Optional[str] = None


class Student(User):
    """Schema for Student model response."""
    admission_number: str
    roll_number: Optional[int] = None
    grade: Optional[str] = None
    section: Optional[str] = None
    admission_date: Optional[date] = None
    date_of_birth: Optional[date] = None
    gender: Optional[str] = None
    blood_group: Optional[str] = None
    nationality: Optional[str] = None
    religion: Optional[str] = None
    address: Optional[str] = None
    city: Optional[str] = None
    county: Optional[str] = None
    country: Optional[str] = None
    whatsapp_number: Optional[str] = None
    emergency_contact: Optional[str] = None
    photo: Optional[str] = None
    status: str
    exit_date: Optional[date] = None
    graduation_date: Optional[date] = None
    withdrawal_reason: Optional[str] = None


class StudentCreateResponse(Student):
    """Schema for student creation response with optional generated password and admission number."""
    generated_password: Optional[str] = None
    generated_admission_number: Optional[str] = None
    
    class Config:
        from_attributes = True

class StudentListResponse(BaseModel):
    """Schema for paginated student list."""
    items: List[Student]
    total: int

class StudentBulkDelete(BaseModel):
    """Schema for bulk student deletion."""
    student_ids: List[UUID]

File: Frontend/sms-client/src/types/student.ts
import { User } from './auth';

export interface Student extends User {
  admission_number: string;
  roll_number?: number;
  grade?: string;
  section?: string;
  admission_date?: string;
  date_of_birth?: string;
  gender?: string;
  blood_group?: string;
  nationality?: string;
  religion?: string;
  address?: string;
  city?: string;
  county?: string;
  country?: string;
  whatsapp_number?: string;
  emergency_contact?: string;
  status: string;
  exit_date?: string;
  graduation_date?: string;
  withdrawal_reason?: string;
  photo?: string;
  phone_number?: string;
  grade_id?: string;
  section_id?: string;
}


export interface StudentCreateResponse extends Student {
  generated_password?: string;
  generated_admission_number?: string;
}
export interface StudentCreate {
  email: string;
  firstName: string;
  lastName: string;
  password?: string;
  admission_number?: string;
  roll_number?: number;
  grade?: string;
  section?: string;
  admission_date?: string;
  date_of_birth?: string;
  gender?: string;
  blood_group?: string;
  nationality?: string;
  religion?: string;
  address?: string;
  city?: string;
  county?: string;
  country?: string;
  whatsapp_number?: string;
  emergency_contact?: string;
  photo?: string;
  phone_number?: string;
}

export interface StudentUpdate {
  firstName?: string;
  lastName?: string;
  email?: string;
  password?: string;
  admission_number?: string;
  roll_number?: number;
  grade?: string;
  section?: string;
  admission_date?: string;
  date_of_birth?: string;
  gender?: string;
  blood_group?: string;
  nationality?: string;
  religion?: string;
  address?: string;
  city?: string;
  county?: string;
  country?: string;
  whatsapp_number?: string;
  emergency_contact?: string;
  photo?: string;
  phone_number?: string;
  status?: string;
  grade_id?: string;
  section_id?: string;
  academic_year?: string;
}
